lte/cloud/go/sms_ll/sms_rp.go

```diff
@@ -0,0 +1,365 @@
+package sms_ll
+
+import (
+	"fmt"
+)
+
+func SMSRPError(field string) error {
+	return fmt.Errorf("smsrp: %s", field)
+}
+
+// Used for both RP-Destination and RP-Originator Address, since it's the
+// same layout (TS 24.011 8.2.5.1 and 8.2.5.2). Note that while 8.2.5.1-2
+// suggest that the first octet is an IEI, 7.3.1 notes that this is a Type
+// 4 LV IE, which means there's no IEI present -- just a length and values.
+type RPAddressElement struct {
+	Length     byte // of the number in half-octets!
+	NumberInfo byte // octet 3
+	Number     []byte
+}
+
+// The length field of an RP Address Element is the number of half-octets in
+// the number. This converts to a byte length of the number itself.
+func (rpadde RPAddressElement) GetNumberOctets() int {
+	l := int(rpadde.Length)
+	if l%2 == 0 {
+		return l/2
+	} else {
+		return (l+1)/2 // last half-octet is padded with 0xf
+	}
+}
+
+func (rpadde RPAddressElement) MarshalBinary() ([]byte, error) {
+	if rpadde.Length == 0x0 {
+		return []byte{rpadde.Length}, nil
+	}
+
+	if int(rpadde.Length) - 1 != len(rpadde.Number) {
+		return nil, SMSRPError("Address size mismatch")
```

Comment:
this validation should be done in `UnmarshalBinary`. You should be able to trust these structs, particularly if you don't export them. ultimately MarshalBinary should not return an error